File: backend/app/schemas/resume.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


# --- Request schemas ---

ROLE_CHOICES = [
    "software_engineer",
    "frontend_engineer",
    "backend_engineer",
    "fullstack_engineer",
    "ai_ml_engineer",
    "data_engineer",
    "devops_sre",
    "mobile_engineer",
    "cybersecurity_engineer",
    "other",
]

TONE_CHOICES = ["conservative", "balanced", "creative"]
TEMPLATE_CHOICES = ["jake", "swe_default"]


class GenerateResumeRequest(BaseModel):
    resume_text: str = Field(..., min_length=10, description="Extracted text from uploaded resume")
    template: str = Field(default="jake", description="Resume template")
    role: str = Field(default="software_engineer", description="Target role")
    job_description: str | None = Field(default=None, description="Optional job description to optimize for")
    tone: str = Field(default="balanced", description="conservative, balanced, or creative")
    github_username: str | None = Field(default=None, description="Optional GitHub username to enhance resume")


class ParseResumeResponse(BaseModel):
    text: str


# --- Resume section schemas ---

class ResumeHeader(BaseModel):
    name: str
    email: str | None = None
    phone: str | None = None
    github_url: str | None = None
    location: str | None = None
    linkedin: str | None = None
    website: str | None = None


class ResumeSkills(BaseModel):
    languages: list[str] = []
    frameworks: list[str] = []
    tools: list[str] = []
    other: list[str] = []


class ResumeProject(BaseModel):
    name: str
    tech_stack: str
    url: str | None = None
    bullets: list[str]


class ResumeExperience(BaseModel):
    title: str
    org: str
    date_range: str | None = None
    bullets: list[str]


class ResumeEducation(BaseModel):
    school: str
    degree: str
    date_range: str | None = None
    details: str | None = None


# --- Response schemas ---

class ResumeResponse(BaseModel):
    id: UUID
    github_username: str | None
    header: ResumeHeader
    summary: str
    skills: ResumeSkills
    projects: list[ResumeProject]
    experience: list[ResumeExperience]
    education: list[ResumeEducation]
    template: str
    is_edited: bool
    credits_remaining: int
    created_at: datetime
    updated_at: datetime


class ResumeUpdateRequest(BaseModel):
    header: ResumeHeader | None = None
    summary: str | None = None
    skills: ResumeSkills | None = None
    projects: list[ResumeProject] | None = None
    experience: list[ResumeExperience] | None = None
    education: list[ResumeEducation] | None = None


class ResumeListItem(BaseModel):
    id: UUID
    github_username: str | None
    header_name: str
    template: str
    is_edited: bool
    created_at: datetime
    updated_at: datetime

File: frontend/src/types/resume.ts
export interface ResumeHeader {
  name: string;
  email: string | null;
  phone: string | null;
  github_url: string | null;
  location: string | null;
  linkedin: string | null;
  website: string | null;
}

export interface ResumeSkills {
  languages: string[];
  frameworks: string[];
  tools: string[];
  other: string[];
}

export interface ResumeProject {
  name: string;
  tech_stack: string;
  url: string | null;
  bullets: string[];
}

export interface ResumeExperience {
  title: string;
  org: string;
  date_range: string | null;
  bullets: string[];
}

export interface ResumeEducation {
  school: string;
  degree: string;
  date_range: string | null;
  details: string | null;
}

export interface ResumeResult {
  id: string;
  github_username: string | null;
  header: ResumeHeader;
  summary: string;
  skills: ResumeSkills;
  projects: ResumeProject[];
  experience: ResumeExperience[];
  education: ResumeEducation[];
  template: string;
  is_edited: boolean;
  credits_remaining: number;
  created_at: string;
  updated_at: string;
}

export interface ResumeListItem {
  id: string;
  github_username: string | null;
  header_name: string;
  template: string;
  is_edited: boolean;
  created_at: string;
  updated_at: string;
}

export interface ATSScoreCategory {
  name: string;
  score: number;
  max_score: number;
  feedback: string;
}

export interface ATSScoreResult {
  overall: number;
  categories: ATSScoreCategory[];
  summary: string;
  recommendations: string[];
}

export interface GenerateResumePayload {
  resume_text: string;
  template: string;
  role: string;
  job_description?: string;
  tone: string;
  github_username?: string;
}

export const ROLE_OPTIONS = [
  { value: "software_engineer", label: "Software Engineer" },
  { value: "frontend_engineer", label: "Frontend Engineer" },
  { value: "backend_engineer", label: "Backend Engineer" },
  { value: "fullstack_engineer", label: "Full Stack Engineer" },
  { value: "ai_ml_engineer", label: "AI / ML Engineer" },
  { value: "data_engineer", label: "Data Engineer" },
  { value: "devops_sre", label: "DevOps / SRE" },
  { value: "mobile_engineer", label: "Mobile Engineer" },
  { value: "cybersecurity_engineer", label: "Cybersecurity Engineer" },
  { value: "other", label: "Other" },
] as const;

export const TONE_OPTIONS = [
  { value: "conservative", label: "Conservative", description: "Minimal changes, professional tone" },
  { value: "balanced", label: "Balanced", description: "Improved clarity with reasonable impact" },
  { value: "creative", label: "Creative", description: "Bold language, maximum impact" },
] as const;

export const TEMPLATE_OPTIONS = [
  { value: "jake", label: "Jake's Resume", description: "Classic LaTeX SWE template" },
  { value: "swe_default", label: "SWE Default", description: "Clean modern template" },
] as const;
